### PR TITLE
prov/efa: Re-post handshake packet failed with EAGAIN previously

### PR DESCRIPTION
**Pull request by zhngaj**
Currently, when using extra feature (e.g., DELIEVERY_COMPLETE),
the sender triggers a handshake procedure, and won't proceed
futher until receiving the handshake packet from the peer.
However, if posting handshake packet fails at peer side, the
sender will keep waiting, which results in hanging issue.

This patch fixes the issue by
1. queuing the peers to which we previously fail to post
    handshake packet with EAGAIN error.
2. re-posting the handshake packet to those queued peers later.

Tested:
1. Add a retry counter (say 2) in rxr_ep structure to intentionally let `rxr_pkt_post_handshake` fail 2 times with -FI_EAGAIN to make sure that peer gets queued in rxr_ep->peer_queued_list and repost 2 times before success.
2. Similar as 1, intentionally let `rxr_pkt_post_handshake` fail the first time (so that queue and repost logic is used), run 128 nodes openfoam with Intel MPI U10, can see that the peer that previously failed with -FI_EAGAIN gets queued and repost later, and openfoam ran into completion. Without this patch, openfoam will hang if we let `rxr_pkt_post_handshake` fail the first time.
3. Fabtests
4. internal PRTest
5. 128 nodes openfoam 80 times back to back.

Signed-off-by: Jie Zhang <zhngaj@amazon.com>

### CODE REVIEW

File: prov/efa/src/rxr/rxr_pkt_cmd.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:883]: I think we need to write eq error here. assert is only effective during debug

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:116]: Consider the possibility of RNR, we need to revisit this logic.
SENT does not mean the packet is received by peer, because it can got RNR. 
I think we need to rename the flag `RXR_PEER_HANDSHAKE_SENT` as `RXR_PEER_HANDSHAKE_SEND_COMPLETE` and set it in `rxr_pkt_handle_handshake_send_completion`

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:74]: We will need to use a `rx_entry` for this. tx_entry pool is fixed sized, and cannot grow.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:82]: -FI_ENOMEM is more appropriate. The caller cannot retry.

**Comment by rwespetal** [comment]
Added do not merge label - Sean we don't plan to merge this fix until after 1.12.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:116]: Actually, I think even for non-RNR case, it's still not right, as resend could also failure. So we should not set this flag before getting completion.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:116]: Yes, that is what I meant. Set the flag in `rxr_pkt_handle_handshake_send_completion`, and rename the flag to RXR_PEER_HANDSHAKE_SEND_COMPLETED

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:116]: offline chatted with @zhngaj. This change is not needed, and can cause multiple handshake being sent.

**Comment by wzamazon** [comment]
I think the title of the PR is a little misleading, we are not queueing the handshake packet. We used the function `rxr_pkt_post_ctrl_or_queue` to post handshake. Which will not queue packet but will release packet if sending failed, and reconstruct the packet each time. I think title should be `use rxr_pkt_post_ctrl_or_queue to send handshake`

**Comment by zhngaj** [comment]
Make sense, we just insert rx/tx_entry->queued_entry into ep's tx/rx_entry_queued_list, and not queue pkt itself.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:103]: Also noted that setting `RXR_PEER_HANDSHAKE_SENT` flag in rxr_pkt_handle_handshake_sent can also cause mutiple handshake being sent, as rxr_pkt_handle_ctrl_sent will not be invoked if rxr_pkt_entry_send returns FI_EAGAIN. 
https://github.com/ofiwg/libfabric/blob/master/prov/efa/src/rxr/rxr_pkt_cmd.c#L339-L348

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:96]: Once -> If

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:541]: Set the flag `RXR_PEER_HANDSHAKE_SENT` here is preferred

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:100]: Why not just call `efa_eq_write_error` here, and do not return error code. This way the caller does not need to change.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:71]: Since you are here, please add a comment describe the function: input, output, behavior. etc

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:541]: @wzamazon I posted the reason why we can not set flag here above.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:541]: Copied here for your reference.
```
Also noted that setting RXR_PEER_HANDSHAKE_SENT flag in rxr_pkt_handle_handshake_sent can also cause mutiple handshake being sent, as rxr_pkt_handle_ctrl_sent will not be invoked if rxr_pkt_entry_send returns FI_EAGAIN.
https://github.com/ofiwg/libfabric/blob/master/prov/efa/src/rxr/rxr_pkt_cmd.c#L339-L348
So I leave it setting in rxr_pkt_post_handshake.
```

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:541]: I see your point. Even if handshake has not been sent, as long as it will be retried, we do not want to send a new handshake. In this case, the flag name `RXR_PKT_HANDSHAKE_SENT` is not accurate. Can you rename it something else like "SENT_OR_QUEUED". Also leave a comment.

**Review by rwespetal — COMMENTED** [review]
> However, if posting handshake packet fails at peer side, the sender will keep waiting, which results in hanging issue.

Was this because no additional packets were received from the sender, so `rxr_pkt_post_handshake` was never called again to send the packet to the sender? Or was it something else?

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: > > However, if posting handshake packet fails at peer side, the sender will keep waiting, which results in hanging issue.
> 
> Was this because no additional packets were received from the sender, so `rxr_pkt_post_handshake` was never called again to send the packet to the sender? Or was it something else?

Correct, the sender sets the flag `RXR_PEER_REQ_SENT` after sending the REQ packet, so the sender just waits for handshake packet and it won't send additonal packets to the peer.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: > I'm not sure if this is the best approach here, now we're going to crash if we fail to allocate an `rx_entry` when this is called. Given we're going to allocate these for every REQ packet received during startup I think that could be a problem.
> 
> I'd prefer something else, maybe a list of peers that we failed to send handshake packets for. Those peer structures are already allocated by the time this function is called. This approach also lets us prioritize sending these packets over others (like segments, etc.) in the progress function.

That's also the reason we switched to use rx_entry pool here, as it can grow.
I agree with you here, and actually thought about something similar. I was thinking to scan all peers in progress engine and retry/repost those peers that we failed to allocate rx_entry for, but felt that's too much cost in the critical path. 
As you sugguested, we probably can add another list struct and just link those failed peer (failed to allocate rx_entry for), and retry/repost when entering progress engine. 

Next step, I plan to address all minor comments in this commit, and have a separate commit in this PR for the peer list that needs to be retried. Let me know if that's acceptable.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: I think use a separate commit to implement peer list is a good idea. We will need to evaluate its performance impact. Adding the peer list means we will be looping over that list in progress engine every time. Even though that list will be mostly empty after initialization, there are still costs associated with looping over it.

Comment (rwespetal) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: Why is the rx_entry needed at all?

The commits should be add peer list + implement this commit, not this commit and then add the peer list, though. And if each commit doesn't stand on its own (compiles and works independently) then just do one.

I'm not worried about checking an empty list and don't think you need to evaluate the impact. I'll take the cache miss for checking two list pointers over trying to obtain an rx_entry in this path.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: In this case `rx_entry` is only used to hold addr, nothing else. If we are not concerned the performance impact, we can do it in two commits. First one introduce queued_peer_list in rxr_ep, and loop over it in progress engine. Next one put the peer in queue_peer_list if posting handshake encountered EAGAIN. In that case, there is not must point of reusing `rxr_pkt_post_ctrl_or_queue`. We might just have `rxr_pkt_post_handshake_or_queue`.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: Alright, I think i got the basic idea here. Once EAGAIN happens when posting handshake packet, we just add that peer into the queue_peer_list and retry later. During retry, we will reconstruct the pkt_entry. So rx_entry is not needed. The work needed is basically as follows:
1. Inside rxr_pkt_handle_recv_completion(), call rxr_pkt_post_handshake_or_queue(). If an error other than EAGAIN returns, hard fail.
2. Inside rxr_pkt_post_handshake_or_queue(), if EAGAIN happens, add that peer into queue_peer_list for retry later.
3. Inside rxr_ep_progress_internal, add a loop to scan queue_peer_list, call rxr_pkt_post_handshake_or_queue for each peer.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: > Inside rxr_ep_progress_internal, add a loop to scan queue_peer_list, call rxr_pkt_post_handshake_or_queue for each peer.

You do not want to call `rxr_pkt_post_handshake_or_queue` inside progress engine, because it is already queued. You need a separate function `rxr_pkt_post_handshake` to just post handshake. Both `rxr_pkt_post_handshake_or_queue` and progress engine call this function.

Comment (rwespetal) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: That's what I had in mind, yeah. Another option is to trigger the handshake during av_insert which could potentially move this traffic to job init/startup time instead of on the first packet. But I think that should be explored separately.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:80]: Okay, then I will put these change in the current commit for simplicity.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c]: Updated in new revision.

File: prov/efa/src/rxr/rxr_pkt_type.h
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type.h:164]: since ```addr``` can be obtained from ```peer``` I do not think you need it in arguments

File: prov/efa/src/rxr/rxr_pkt_type.h
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type.h:160]: For constituency, maybe use ```peer``` as argument?

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:115]: use /*  */for comment for consistency.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:116]: Is `queued_entry` used to link to lists other than `ep->queued_peer_list`?

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:116]: No, it's used to list peer to `ep->queued_peer_list`.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:87]: Now we queue the peer, so return -FI_EAGAIN in order to retry.

**Comment by zhngaj** [comment]
travis-ci failure is not related to the change in this commit. 
```
name:   fi_eq_test -p "udp"
  result: Fail
  time:   2
  server_cmd: fi_eq_test -p "udp"
  server_stdout: |
    Testing EQs on fabric fe80::/64
    Running eq_open_close [Test open and close of EQ for various sizes]...PASS!
    Running eq_write_read_self [Test writing and reading to EQ]...PASS!
    Running eq_size_verify [Test EQ supports writing # of entries = size]...PASS!
    Running eq_wait_fd_poll [Test polling on fd extracted from EQ]...skipped because: fi_control ret=-96, No message available on STREAM
    Running eq_wait_fd_sread [Test EQ sread]...FAIL: fi_eq_sread slept 9 ms, expected immediate return
    Running eq_wait_read_peek [Test EQ read with FI_PEEK]...PASS!
    Running eq_wait_sread_peek [Test EQ sread with FI_PEEK]...PASS!
    Running eq_readerr_ret_eagain [Test EQ readerr with FI_EAGAIN]...PASS!
```

File: prov/efa/src/rxr/rxr_pkt_cmd.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:678]: I think you removed it by mistake.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_cmd.c:678]: The `break` here is not needed, as the next `case RXR_CTS_PKT` will break anyway. 
It's not necessary to open another PR fixing this, so remove it here.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.c:678]: Rely on fall back in switch case is not good practice. What if in the future we add some special handling for CTS packet? say call function rxr_pkt_handle_cts_send_completion()?

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_cmd.c:678]: Alright, I will change it back.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_pkt_cmd.c]: Fixed in new rev.

**Review by wzamazon — APPROVED** [review]

File: prov/efa/src/rxr/rxr.h
Comment (rwespetal) [inline prov/efa/src/rxr/rxr.h:656]: s/handshat/handshake

File: prov/efa/src/rxr/rxr_ep.c
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_ep.c:1617]: This doesn't match with the list you're iterating as the peers are queued not the packets. Perhaps change to

"Resend handshake packet for any peers where the first handshake send failed."

File: prov/efa/src/rxr/rxr_ep.c
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_ep.c:1626]: Pick one style here (I prefer no braces if one line only).

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:75]: Please follow doxygen style for function header comments.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:72]: L72-73 isn't needed.

**Review by rwespetal — APPROVED** [review]
Just some minor nits, looks good to me.

**Review by wzamazon — APPROVED** [review]

**Pull request merged**
